src/Control/Newtype/Generic.hs

```diff
@@ -0,0 +1,210 @@
+{-# language TypeOperators              #-}
+{-# language DataKinds                  #-}
+{-# language UndecidableInstances       #-}
+
+{- |
+A version of the 'Newtype' typeclass and related functions.
+Primarily pulled from Conor McBride's Epigram work. Some examples:
+
+>>> ala Sum foldMap [1,2,3,4]
+10
+
+>>> ala Endo foldMap [(+1), (+2), (subtract 1), (*2)] 3
+8
+
+>>> under2 Min (<>) 2 1
+1
+
+>>> over All not (All False)
+All {getAll = True)
+
+The version of the 'Newtype' class exported by this module has only
+one instance, which cannot be overridden. It has instances for
+*all* and *only* newtypes with 'Generic' instances whose generated
+'Coercible' instances are visible. This makes it more convenient than the
+version in "Control.Newtype.Generics" when it applies.
+
+Like McBride's version, and unlike the one in "Control.Newtype.Generics",
+this version has two parameters: one for the newtype and one for the
+underlying type. This generally makes 'Newtype' constraints more compact
+and easier to read.
+
+Note: Each function in this module takes an argument representing a
+newtype constructor. This is used only for its type. To make that clear,
+the type of that argument is allowed to be extremely polymorphic: @o `to` n@
+rather than @o -> n@.
+
+@since TODO
+-}
+module Control.Newtype.Generic
+  ( Newtype
+  , O
+  , op
+  , ala
+  , ala'
+  , under
+  , over
+  , under2
+  , over2
+  , underF
+  , overF
+  ) where
+
+import GHC.Generics
+import Data.Coerce
+#if MIN_VERSION_base(4,9,0)
+import GHC.TypeLits (TypeError, ErrorMessage (..))
+#endif
+
+-- | Get the underlying type of a newtype.
+--
+-- @
+-- data N = N Int deriving Generic
+-- -- O N = Int
+-- @
+type O x = GO x (Rep x)
+
+type family GO x rep where
+  GO _x (D1 ('MetaData _n _m _p 'True) (C1 _c (S1 _s (K1 _i a)))) = a
+#if MIN_VERSION_base(4,9,0)
+  GO x _rep = TypeError
+    ('Text "There is no " ':<>: 'ShowType Newtype ':<>: 'Text " instance for"
+      ':$$: 'Text "    " ':<>: 'ShowType x
+      ':$$: 'Text "because it is not a newtype.")
+#endif
+
+-- | @Newtype n o@ means that @n@ is a newtype wrapper around
+-- @o@. @n@ must be an instance of 'Generic'. Furthermore, the
+-- @'Coercible' n o@ instance must be visible; this typically
+-- means the newtype constructor is visible, but the instance
+-- could also have been brought into view by pattern matching
+-- on a 'Data.Type.Coercion.Coercion'.
+class (Coercible n o, O n ~ o) => Newtype n o
+
+-- The Generic n constraint gives a much better type error if
+-- n is not an instance of Generic. Without that, there's
+-- just a mysterious message involving GO and Rep. With it, the
+-- lousy error message still shows up, but at least there's
+-- also a good one.
+instance (Generic n, Coercible n o, O n ~ o) => Newtype n o
+
+-- Compat shim for GHC 7.8
+coerce' :: Coercible a b => b -> a
+#if __GLASGOW_HASKELL__ >= 710
+coerce' = coerce
+#else
+coerce' = coerce (\x -> x :: a) :: forall a b. Coercible a b => b -> a
+#endif
```

Comment:
I see the rest of the package doesn't support GHC 7.8. If that's going to continue to be the case, then I can remove this shim.